MSAL/src/native_auth/controllers/sign_in/MSALNativeAuthSignInController.swift

```diff
@@ -358,13 +319,222 @@ final class MSALNativeAuthSignInController: MSALNativeAuthTokenController, MSALN
+                    self?.stopTelemetryEvent(event, context: context, delegateDispatcherResult: result)
+                })
+        case .introspectRequired:
+            let telemetryInfo = TelemetryInfo(event: event, context: context)
+            let response = await performAndValidateIntrospectRequest(continuationToken: continuationToken, context: context)
+            let introspectResponse = handleIntrospectResponse(
+                response, scopes: scopes,
+                telemetryInfo: telemetryInfo,
+                continuationToken: continuationToken
+            )
+            switch introspectResponse.result {
+            case .selectionRequired(let authMethods, let newState):
+                return .init(.selectionRequired(
+                    authMethods: authMethods,
+                    newState: newState),
+                             correlationId: introspectResponse.correlationId,
+                             telemetryUpdate: { [weak self] result in
+                                 self?.stopTelemetryEvent(telemetryInfo.event, context: telemetryInfo.context, delegateDispatcherResult: result)
+                             })
+            case .error(let error, let newState):
+                return .init(.error(error: error, newState: newState), correlationId: introspectResponse.correlationId)
+            }
+        }
+    }
+
+    func getAuthMethods(
+        continuationToken: String,
+        context: MSALNativeAuthRequestContext,
+        scopes: [String]
+    ) async -> MFAGetAuthMethodsControllerResponse {
+        let event = makeAndStartTelemetryEvent(id: .telemetryApiIdMFAGetAuthMethods, context: context)
+        let result = await performAndValidateIntrospectRequest(continuationToken: continuationToken, context: context)
+        let telemetryInfo = TelemetryInfo(event: event, context: context)
+        return handleIntrospectResponse(result, scopes: scopes, telemetryInfo: telemetryInfo, continuationToken: continuationToken)
+    }
+
+    func submitChallenge(
+        challenge: String,
+        continuationToken: String,
+        context: MSALNativeAuthRequestContext,
+        scopes: [String]
+    ) async -> MFASubmitChallengeControllerResponse {
+        let event = makeAndStartTelemetryEvent(id: .telemetryApiIdMFASubmitChallenge, context: context)
+        let response = await submitCode(
+            challenge,
+            continuationToken: continuationToken,
+            context: context,
+            scopes: scopes,
+            telemetryEvent: event
+        )
+        switch response.result {
+        case .completed(let accountResult):
+            return .init(
+                .completed(accountResult),
+                correlationId: response.correlationId,
+                telemetryUpdate: { [weak self] result in
+                    self?.stopTelemetryEvent(event, context: context, delegateDispatcherResult: result)
+                })
+        case .error(let error, let newState):
```

Comment:
this calls the /token endpoint, right? Where is the handling for the API returning mfa_required again? I know it can't happen practically right now, but it can happen in the future. Will that response result in an .error case?

Reply:
In the `MSALNativeAuthTokenResponseValidator` class the "mfa_required" is converted to a dedicated internal validated result that is "strongAuthRequired". This type of result will be converted to an error for the following scenario:
- signIn after signUp/SSPR
- signIn with username and code
Only during the signIn with password this result is converted to the state "awaitingMFA" .

Reply:
That's not exactly my point. What I mean is: when you submit a challenge during the sign in with password flow, i.e. in the MFA flow, and the API *again* returns mfa_required (e.g. if another MFA cycle is needed), what happens?

Reply:
Ah, now I got you. We will return an error because we don't expect a strong authentication after submitting a code (second factor after signIn with username and code and after submitting a challenge). Check line 531